plotly/widgets/graph_widget.py

```diff
@@ -0,0 +1,130 @@
+        # iframe to communicate that it is ready
+        # unfortunately, this two-way blocking communication
+        # isn't possible (https://github.com/ipython/ipython/wiki/IPEP-21:-Widget-Messages#caveats)
+        # so we'll just cue up messages until they're ready to be sent
+        self._clientMessages = deque()
+
+    def _handle_msg(self, message):
+        """Handle a msg from the front-end.
+        Parameters
+        ----------
+        content: dict
+            Content of the msg."""
+        content = message['content']['data']['content']
+        if content.get('event', '') == 'pong':
+            self._graphId = content['graphId']
+
+            # ready to recieve - pop out all of the items in the deque
+            while self._clientMessages:
+                _message = self._clientMessages.popleft()
+                _message['graphId'] = self._graphId
+                _message = json.dumps(_message)
+                self._message = _message
+
+        if content.get('event', '') in ['click', 'hover', 'zoom']:
+            self._event_handlers[content['event']](self, content)
+
+    def _handle_registration(self, event_type, callback, remove):
+        self._event_handlers[event_type].register_callback(callback,
+                                                           remove=remove)
+        event_callbacks = self._event_handlers[event_type].callbacks
+        if (len(event_callbacks) and event_type not in self._listener_set):
+            self._listener_set.add(event_type)
+            message = {'listen': list(self._listener_set)}
+            self._handle_outgoing_message(message)
+
+    def _handle_outgoing_message(self, message):
+        if self._graphId == '':
+            self._clientMessages.append(message)
+        else:
+            message['graphId'] = self._graphId
+            self._message = json.dumps(message)
+
+    def on_click(self, callback, remove=False):
+        """Register a callback to execute when the graph is clicked.
+        Parameters
+        ----------
+        remove : bool (optional)
+            Set to true to remove the callback from the list of callbacks."""
+        self._handle_registration('click', callback, remove)
+
+    def on_hover(self, callback, remove=False):
+        """Register a callback to execute when you hover over points in the graph.
+        Parameters
+        ----------
+        remove : bool (optional)
+            Set to true to remove the callback from the list of callbacks."""
+        self._handle_registration('hover', callback, remove)
+
+    def on_zoom(self, callback, remove=False):
+        """Register a callback to execute when you zoom in the graph.
+        Parameters
+        ----------
+        remove : bool (optional)
+            Set to true to remove the callback from the list of callbacks."""
+        self._handle_registration('zoom', callback, remove)
+
+    def restyle(self, data, traces=None):
```

Comment:
I'd vote for making `restyle` and `relayout` private and adding an `update` public method which would call restyle / relayout depending on the nature of the data sent.